File: biglinux-webapps/usr/share/biglinux/webapps/webapps/ui/template_gallery.py
"""
Template gallery dialog — browse and apply curated webapp templates.
"""

import gi

gi.require_version("Gtk", "4.0")
gi.require_version("Adw", "1")
from gi.repository import Gtk, Adw, GObject

from webapps.templates.registry import WebAppTemplate, build_default_registry
from webapps.utils.translation import _

import logging

logger = logging.getLogger(__name__)


class TemplateGallery(Adw.Window):
    """Grid gallery of curated webapp templates."""

    __gsignals__ = {
        "template-selected": (
            GObject.SignalFlags.RUN_FIRST,
            None,
            (str,),  # template_id
        ),
    }

    def __init__(self, parent: Gtk.Window) -> None:
        super().__init__(
            title=_("Templates"),
            transient_for=parent,
            modal=True,
            destroy_with_parent=True,
            default_width=650,
            default_height=500,
        )

        self.registry = build_default_registry()
        self._build_ui()

    # ── UI ──────────────────────────────────────────────────────────

    def _build_ui(self) -> None:
        content = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=0)

        # header
        header = Adw.HeaderBar()
        header.set_title_widget(Gtk.Label(label=_("Choose a Template")))
        header.add_css_class("flat")
        content.append(header)

        # search bar
        self.search_entry = Gtk.SearchEntry()
        self.search_entry.set_placeholder_text(_("Search templates..."))
        self.search_entry.set_margin_start(12)
        self.search_entry.set_margin_end(12)
        self.search_entry.set_margin_top(6)
        self.search_entry.set_margin_bottom(6)
        self.search_entry.update_property(
            [Gtk.AccessibleProperty.LABEL],
            [_("Search templates")],
        )
        self.search_entry.connect("search-changed", self._on_search_changed)
        content.append(self.search_entry)

        # scrolled area
        scrolled = Gtk.ScrolledWindow()
        scrolled.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)
        scrolled.set_vexpand(True)

        self.main_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=0)
        scrolled.set_child(self.main_box)
        content.append(scrolled)

        self._populate_all()
        self.set_content(content)

    def _populate_all(self) -> None:
        """Populate gallery with all templates grouped by category."""
        self._clear_main_box()
        categories = self.registry.get_categories()
        for cat in sorted(categories):
            templates = self.registry.get_by_category(cat)
            if templates:
                self._add_category_section(cat, templates)

    def _populate_search(self, query: str) -> None:
        """Populate gallery with search results."""
        self._clear_main_box()
        results = self.registry.search(query)
        if results:
            self._add_category_section(_("Search Results"), results)
        else:
            label = Gtk.Label(label=_("No templates found"))
            label.set_margin_top(24)
            label.add_css_class("dim-label")
            self.main_box.append(label)

    def _add_category_section(
        self, title: str, templates: list[WebAppTemplate]
    ) -> None:
        """Add a category header + FlowBox of template cards."""
        # category heading
        heading = Gtk.Label(label=title, xalign=0)
        heading.add_css_class("heading")
        heading.set_margin_start(16)
        heading.set_margin_top(12)
        heading.set_margin_bottom(4)
        self.main_box.append(heading)

        flow = Gtk.FlowBox()
        flow.set_selection_mode(Gtk.SelectionMode.NONE)
        flow.set_homogeneous(True)
        flow.set_max_children_per_line(5)
        flow.set_min_children_per_line(2)
        flow.set_margin_start(12)
        flow.set_margin_end(12)
        flow.set_margin_bottom(8)
        flow.set_row_spacing(8)
        flow.set_column_spacing(8)

        for tmpl in templates:
            card = self._make_card(tmpl)
            flow.append(card)

        self.main_box.append(flow)

    def _make_card(self, tmpl: WebAppTemplate) -> Gtk.Button:
        """Build a clickable card for a template."""
        box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=4)
        box.set_margin_top(8)
        box.set_margin_bottom(8)
        box.set_margin_start(4)
        box.set_margin_end(4)
        box.set_halign(Gtk.Align.CENTER)

        icon = Gtk.Image.new_from_icon_name(tmpl.icon or "application-x-executable")
        icon.set_pixel_size(48)
        box.append(icon)

        label = Gtk.Label(label=tmpl.name)
        label.set_ellipsize(3)  # PANGO_ELLIPSIZE_END
        label.set_max_width_chars(14)
        label.add_css_class("caption")
        box.append(label)

        btn = Gtk.Button()
        btn.set_child(box)
        btn.add_css_class("flat")
        btn.set_tooltip_text(tmpl.comment or tmpl.name)
        btn.connect("clicked", self._on_card_clicked, tmpl.template_id)

        btn.update_property(
            [Gtk.AccessibleProperty.LABEL],
            [tmpl.name],
        )

        return btn

    def _clear_main_box(self) -> None:
        while child := self.main_box.get_first_child():
            self.main_box.remove(child)

    # ── Signals ─────────────────────────────────────────────────────

    def _on_search_changed(self, entry: Gtk.SearchEntry) -> None:
        query = entry.get_text().strip()
        if query:
            self._populate_search(query)
        else:
            self._populate_all()

    def _on_card_clicked(self, _btn: Gtk.Button, template_id: str) -> None:
        self.emit("template-selected", template_id)
        self.close()

File: crates/webapps-manager/src/template_gallery.rs
use gtk4 as gtk;
use libadwaita as adw;

use crate::platform::tooltip;
use adw::prelude::*;
use gettextrs::gettext;
use std::cell::RefCell;
use std::rc::Rc;
use webapps_core::templates::{default_registry, TemplateRegistry, WebAppTemplate};

/// One-shot owning callback container — fires its inner closure at most once.
type OnceCallbackCell = Rc<RefCell<Option<Box<dyn FnOnce(String)>>>>;

/// Show template gallery. Fires the callback at most once on the first
/// selection — guards against the user double-activating a row before the
/// modal closes, which would otherwise apply the template twice and clobber
/// any edits made between the two invocations.
pub fn show(parent: &impl IsA<gtk::Widget>, on_selected: impl FnOnce(String) + 'static) {
    let registry = default_registry();
    let on_selected_cell: OnceCallbackCell = Rc::new(RefCell::new(Some(Box::new(on_selected))));
    let callback: Rc<dyn Fn(String)> = {
        let cell = on_selected_cell.clone();
        Rc::new(move |template_id| {
            if let Some(cb) = cell.borrow_mut().take() {
                cb(template_id);
            }
        })
    };

    let dialog = adw::Dialog::builder()
        .title(gettext("Choose a Template"))
        .build();
    crate::geometry::bind_adw_dialog(&dialog, "template-gallery.json", 600, 500);

    let toolbar = adw::ToolbarView::new();
    let header = adw::HeaderBar::new();
    let search_entry = gtk::SearchEntry::new();
    search_entry.set_placeholder_text(Some(&gettext("Search templates...")));
    search_entry.set_hexpand(true);
    search_entry.update_property(&[gtk::accessible::Property::Label(&gettext(
        "Search templates",
    ))]);
    header.set_title_widget(Some(&search_entry));
    toolbar.add_top_bar(&header);

    let scroll = gtk::ScrolledWindow::new();
    scroll.set_vexpand(true);
    scroll.set_hscrollbar_policy(gtk::PolicyType::Never);

    let main_box = gtk::Box::new(gtk::Orientation::Vertical, 16);
    main_box.set_margin_top(12);
    main_box.set_margin_bottom(12);
    main_box.set_margin_start(12);
    main_box.set_margin_end(12);

    scroll.set_child(Some(&main_box));
    toolbar.set_content(Some(&scroll));
    dialog.set_child(Some(&toolbar));

    populate_all(&main_box, registry, &callback, &dialog);

    {
        let mb = main_box.clone();
        let cb = callback.clone();
        let d = dialog.clone();
        search_entry.connect_search_changed(move |entry| {
            let query = entry.text().to_string();
            clear_box(&mb);
            if query.is_empty() {
                populate_all(&mb, registry, &cb, &d);
            } else {
                populate_search(&mb, registry, &query, &cb, &d);
            }
        });
    }

    dialog.present(Some(parent));
}

fn populate_all(
    container: &gtk::Box,
    registry: &TemplateRegistry,
    callback: &Rc<dyn Fn(String)>,
    dialog: &adw::Dialog,
) {
    let mut categories = registry.categories();
    categories.sort();
    for cat in &categories {
        let templates = registry.get_by_category(cat);
        if templates.is_empty() {
            continue;
        }
        add_category_section(container, cat, &templates, callback, dialog);
    }
}

fn populate_search(
    container: &gtk::Box,
    registry: &TemplateRegistry,
    query: &str,
    callback: &Rc<dyn Fn(String)>,
    dialog: &adw::Dialog,
) {
    let results = registry.search(query);
    if results.is_empty() {
        let label = gtk::Label::new(Some(&gettext("No templates found")));
        label.add_css_class("dim-label");
        label.set_margin_top(24);
        container.append(&label);
        return;
    }
    add_category_section(
        container,
        &gettext("Search Results"),
        &results,
        callback,
        dialog,
    );
}

fn add_category_section(
    container: &gtk::Box,
    category: &str,
    templates: &[&WebAppTemplate],
    callback: &Rc<dyn Fn(String)>,
    dialog: &adw::Dialog,
) {
    let header = gtk::Label::new(Some(category));
    header.set_halign(gtk::Align::Start);
    header.add_css_class("title-4");
    header.set_margin_top(8);
    header.set_accessible_role(gtk::AccessibleRole::Heading);
    container.append(&header);

    let flow = gtk::FlowBox::new();
    flow.set_selection_mode(gtk::SelectionMode::None);
    flow.set_homogeneous(true);
    flow.set_row_spacing(6);
    flow.set_column_spacing(6);
    flow.set_min_children_per_line(2);
    flow.set_max_children_per_line(4);
    flow.update_property(&[gtk::accessible::Property::Label(category)]);

    for tpl in templates {
        flow.append(&build_template_tile(tpl, callback, dialog));
    }

    container.append(&flow);
}

/// One template rendered as an icon tile.
///
/// The activatable widget is a real `gtk::Button` (not a bare activatable
/// `FlowBoxChild`) so it is keyboard-reachable and AT-SPI exposes a click
/// action — a `FlowBoxChild` alone would report `actions=[]`.
fn build_template_tile(
    tpl: &WebAppTemplate,
    callback: &Rc<dyn Fn(String)>,
    dialog: &adw::Dialog,
) -> gtk::FlowBoxChild {
    let content = gtk::Box::new(gtk::Orientation::Vertical, 6);
    content.set_halign(gtk::Align::Center);
    content.set_margin_top(10);
    content.set_margin_bottom(10);
    content.set_margin_start(6);
    content.set_margin_end(6);
    content.set_width_request(108);

    let icon = gtk::Image::new();
    icon.set_pixel_size(48);
    icon.set_accessible_role(gtk::AccessibleRole::Presentation);
    crate::webapp_row::load_icon(&icon, &tpl.icon);
    content.append(&icon);

    let name = gtk::Label::new(Some(&tpl.name));
    name.set_wrap(true);
    name.set_justify(gtk::Justification::Center);
    name.set_max_width_chars(12);
    name.add_css_class("caption");
    content.append(&name);

    // DRM templates can only run in an external browser. Tooltips aren't
    // announced by AT-SPI, so the constraint goes into the accessible name too.
    let mut accessible_name = tpl.name.clone();
    if tpl.requires_drm {
        let drm_hint = gettext("Requires external browser (DRM)");
        let drm_icon = gtk::Image::from_icon_name("web-browser-symbolic");
        drm_icon.set_pixel_size(12);
        drm_icon.set_accessible_role(gtk::AccessibleRole::Presentation);
        drm_icon.add_css_class("dim-label");
        tooltip::set(&drm_icon, &drm_hint);
        content.append(&drm_icon);
        accessible_name = format!("{} — {drm_hint}", tpl.name);
    }

    let button = gtk::Button::builder().child(&content).build();
    button.add_css_class("flat");
    tooltip::set(&button, &tpl.url);
    button.update_property(&[gtk::accessible::Property::Label(&accessible_name)]);

    let cb = callback.clone();
    let tid = tpl.template_id.clone();
    let d = dialog.clone();
    button.connect_clicked(move |_| {
        cb(tid.clone());
        d.close();
    });

    let child = gtk::FlowBoxChild::new();
    child.set_child(Some(&button));
    // Focus lives on the inner button so keyboard traversal hits the action.
    child.set_focusable(false);
    child
}

fn clear_box(bx: &gtk::Box) {
    while let Some(child) = bx.first_child() {
        bx.remove(&child);
    }
}
